### PR TITLE
[TAS-2060] ✨ Retrieve JWT token during wallet connection

### PR DESCRIPTION
**Pull request by AuroraHuang22**

### CODE REVIEW

File: store/wallet.ts [hunk 1]
```diff
@@ -220,13 +221,63 @@ export default class Wallet extends VuexModule {
...
+        gas: '0',
+        amount: [{
+            denom: LIKECOIN_WALLET_CONNECTOR_CONFIG.coinDenom,
+            amount: '0',
+          }],
+      },
+      sequence: '0',
+      account_number: '0',
+    }
+    if ('signAmino' in this.signer) {
```
Comment (williamchong) [inline store/wallet.ts:254]: Consider also supporting `signArbitrary`

File: components/AppHeader.vue [hunk 2]
```diff
@@ -157,14 +157,21 @@ import logTrackerEvent, { setLoggerUser } from '~/utils/logger'
...
+const bookApiModule = namespace('book-api')
 
 @Component
 export default class AppHeader extends Vue {
   @walletModule.Getter('getType') walletType!: string | null
   @walletModule.Action('disconnectWallet') disconnectWallet!: () => void
   @walletModule.Action('openConnectWalletModal') openConnectWalletModal!: (params: { language: string, fullPath?: string }) => Promise<any>
   @walletModule.Action('initWallet') initWallet!: (params: { method: any, accounts: any, offlineSigner?: any }) => Promise<any>
+  @walletModule.Action('signMessageMemo') signMessageMemo!: (action: string, permissions?: string[]) => Promise<any>
   @walletModule.Getter('getWalletAddress') currentAddress!: string
```
Comment (williamchong) [inline components/AppHeader.vue:169]: Please check where do we use `getWalletAddress `, and consider replacing them all with `sessionWallet`

File: components/AppHeader.vue [hunk 3]
```diff
@@ -201,9 +210,25 @@ export default class AppHeader extends Vue {
         wallet: accounts[0].address,
         method,
       })
-      return this.initWallet(connection)
+      await this.initWallet(connection)
+      if (!this.currentAddress || !this.signer) return
```
Comment (williamchong) [inline components/AppHeader.vue:214]: `isLoading` is not set back to false in all `return` cases, consider using `finally`

File: store/wallet.ts [hunk 4]
```diff
@@ -220,13 +221,63 @@ export default class Wallet extends VuexModule {
...
+        this.address,
+        signingPayload,
+      )
+      return {
+        signature: signature.signature,
+        publicKey: signature.pub_key.value,
+        message: stringify(signed),
+        wallet: this.address,
+        signMethod: 'memo',
+        expiresIn: '1d',
```
Comment (williamchong) [inline store/wallet.ts:265]: Prefer `7d` or `30d`

File: store/book-api.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    this.isRestoringSession = inputIsRestoringSession
+  }
+
+  @Action
+  async authenticate({inputWallet = '', signature = {}}: {inputWallet?: string, signature?: any}) {
+    try {
+      const { data } = await axios.post(
+        API_POST_AUTHORIZE,
+        {
+          expiresIn: '7d',
```
Comment (williamchong) [inline store/book-api.ts:40]: Why set `expiresIn` here when it is also set in `signMessageMemo` ?

**Review by williamchong — CHANGES_REQUESTED** [review]

File: store/wallet.ts (hunk 1 above)
Comment (AuroraHuang22) [reply store/wallet.ts:287]: [✨ Supporting signArbitrary](https://github.com/likecoin/app-like-co/pull/474/commits/181c72d913cd054927d73886af8ae1c340677750)

File: components/AppHeader.vue (hunk 2 above)
Comment (AuroraHuang22) [reply components/AppHeader.vue:175]: [🚸 Improve login UX flow](https://github.com/likecoin/app-like-co/pull/474/commits/e6861c5e5254346b34d2cc1adef7a12f1f2ae05c)

File: store/wallet.ts (hunk 4 above)
Comment (AuroraHuang22) [reply store/wallet.ts]: [✨ Supporting signArbitrary](https://github.com/likecoin/app-like-co/pull/474/commits/181c72d913cd054927d73886af8ae1c340677750)

File: store/book-api.ts (hunk 5 above)
Comment (AuroraHuang22) [reply store/book-api.ts]: [🎨 Remove duplicate expiresIn](https://github.com/likecoin/app-like-co/pull/474/commits/1015cedfc26cdcf0cd51c3bd319f4994d2b0e3a1)

File: components/AppHeader.vue [hunk 6]
```diff
@@ -104,7 +110,7 @@
                     'overflow-hidden',
                     'overflow-ellipsis',
                   ]"
-                >{{ currentAddress }}</div>
+                >{{ sessionWallet }}</div>
```
Comment (williamchong) [inline components/AppHeader.vue:113]: But it seems that I can still read `My works` and contents in each work before I sign the message
e.g.
image
image

Can we only consider a user is connected after they properly sign a message and has a valid token?

**Review by williamchong — CHANGES_REQUESTED** [review]

File: store/book-api.ts [hunk 7]
```diff
@@ -37,7 +37,6 @@ export default class BookAPI extends VuexModule {
       const { data } = await axios.post(
         API_POST_AUTHORIZE,
         {
-          expiresIn: '7d',
           ...signature,
         },
```
[the remark this answers is not in the document]
Comment (AuroraHuang22) [reply store/book-api.ts]: [♻️ Simplify post format](https://github.com/likecoin/app-like-co/pull/474/commits/55af157390a17cbddf8f7b22b396d230936672d4)

File: components/AppHeader.vue (hunk 6 above)
Comment (AuroraHuang22) [reply components/AppHeader.vue:113]: [👔 Use session wallet to determine connected user](https://github.com/likecoin/app-like-co/pull/474/commits/57d3ef7ba6f1da73bb5519203359a8b310113637)

File: components/AppHeader.vue [hunk 8]
```diff
@@ -201,9 +217,27 @@ export default class AppHeader extends Vue {
         wallet: accounts[0].address,
         method,
       })
-      return this.initWallet(connection)
+      await this.initWallet(connection)
+      if (!this.currentAddress || !this.signer) return
```
Comment (williamchong) [inline components/AppHeader.vue:221]: Throw some error here for ui display?

File: components/AppHeader.vue [hunk 9]
```diff
@@ -201,9 +217,27 @@ export default class AppHeader extends Vue {
...
+      const signature = await this.signMessageMemo('authorize', [
+        'profile',
+        'read:nftbook',
+        'write:nftbook',
+        'read:nftcollection',
+        'write:nftcollection',
+      ])
+      if (!signature) { return }
```
Comment (williamchong) [inline components/AppHeader.vue:229]: Maybe throw `SIGNING_REJECTED` error

File: layouts/wallet.vue [hunk 10]
```diff
@@ -17,51 +25,87 @@ import logTrackerEvent, { setLoggerUser } from '~/utils/logger'
...
+          })
+          await this.initWallet(connection)
+          if (!this.walletAddress || !this.signer) return
+          const signature = await this.signMessageMemo('authorize', [
+            'profile',
+            'read:nftbook',
+            'write:nftbook',
+            'read:nftcollection',
+            'write:nftcollection',
+          ])
```
Comment (williamchong) [inline layouts/wallet.vue:87]: Maybe refactor to a constant for easier maintainance

File: layouts/wallet.vue [hunk 11]
```diff
@@ -17,51 +25,87 @@ import logTrackerEvent, { setLoggerUser } from '~/utils/logger'
...
+            `connected_wallet_${method}`,
+            'connected_wallet',
+            1,
+          )
+          setLoggerUser(this, {
+            wallet: accounts[0].address,
+            method,
+          })
+          await this.initWallet(connection)
+          if (!this.walletAddress || !this.signer) return
```
Comment (williamchong) [inline layouts/wallet.vue:80]: Prefer throwing error for ui display

File: layouts/wallet.vue [hunk 12]
```diff
@@ -17,51 +25,87 @@ import logTrackerEvent, { setLoggerUser } from '~/utils/logger'
...
+          const signature = await this.signMessageMemo('authorize', [
+            'profile',
+            'read:nftbook',
+            'write:nftbook',
+            'read:nftcollection',
+            'write:nftcollection',
+          ])
+          if (!signature) {
+            return
```
Comment (williamchong) [inline layouts/wallet.vue:89]: Prefer throwing error for ui display

File: components/AppHeader.vue [hunk 13]
```diff
@@ -201,9 +217,27 @@ export default class AppHeader extends Vue {
...
       })
-      return this.initWallet(connection)
+      await this.initWallet(connection)
+      if (!this.currentAddress || !this.signer) return
+      const signature = await this.signMessageMemo('authorize', [
+        'profile',
+        'read:nftbook',
+        'write:nftbook',
+        'read:nftcollection',
+        'write:nftcollection',
```
Comment (williamchong) [inline components/AppHeader.vue:227]: Maybe refactor to a constant for easier maintainance

File: pages/auth/redirect.vue [hunk 14]
```diff
@@ -41,6 +49,16 @@ export default class RedirectPage extends Vue {
             method: method as string,
           })
           await this.initWallet(connection)
+          if (!this.currentAddress || !this.signer) return
+          const signature = await this.signMessageMemo('authorize', [
+            'profile',
+            'read:nftbook',
+            'write:nftbook',
+            'read:nftcollection',
+            'write:nftcollection',
```
Comment (williamchong) [inline pages/auth/redirect.vue:58]: Maybe refactor to a constant for easier maintainance

File: pages/auth/redirect.vue [hunk 15]
```diff
@@ -41,6 +49,16 @@ export default class RedirectPage extends Vue {
             method: method as string,
           })
           await this.initWallet(connection)
+          if (!this.currentAddress || !this.signer) return
```
Comment (williamchong) [inline pages/auth/redirect.vue:52]: Prefer throwing error for ui display

File: pages/auth/redirect.vue [hunk 16]
```diff
@@ -41,6 +49,16 @@ export default class RedirectPage extends Vue {
...
+          const signature = await this.signMessageMemo('authorize', [
+            'profile',
+            'read:nftbook',
+            'write:nftbook',
+            'read:nftcollection',
+            'write:nftcollection',
+          ])
+          if (!signature) { return }
```
Comment (williamchong) [inline pages/auth/redirect.vue:60]: Prefer throwing error for ui display

File: store/wallet.ts [hunk 17]
```diff
@@ -220,13 +220,97 @@ export default class Wallet extends VuexModule {
     return connection;
   }
 
+  @Action
+  // eslint-disable-next-line class-methods-use-this
```
Comment (williamchong) [inline store/wallet.ts:224]: This function indeed used `this`, why do we need this comment?

File: store/book-api.ts [hunk 18]
```diff
@@ -0,0 +1,90 @@
...
+      if (!token) {
+        throw new Error('INVALID_SIGNATURE')
+      }
+      this.context.commit('setToken', token)
+      this.context.commit('setSessionWallet', inputWallet)
+      saveAuthSession({ wallet: inputWallet, token})
+    } catch (error) {
+      // eslint-disable-next-line no-console
+      console.error(error)
+      throw new Error('AUTHENTICATION_FAILED')
```
Comment (williamchong) [inline store/book-api.ts:51]: Maybe include the http body in the error message throw, if the error is an axios response

**Review by williamchong — CHANGES_REQUESTED** [review]

File: pages/auth/redirect.vue (hunk 16 above)
Comment (AuroraHuang22) [reply pages/auth/redirect.vue]: [🥅 Throw connection error](https://github.com/likecoin/app-like-co/pull/474/commits/cc9826077540dfdcd12ebe0f57470e7419f8c922)

File: pages/auth/redirect.vue (hunk 15 above)
Comment (AuroraHuang22) [reply pages/auth/redirect.vue]: [🥅 Throw connection error](https://github.com/likecoin/app-like-co/pull/474/commits/cc9826077540dfdcd12ebe0f57470e7419f8c922)

File: layouts/wallet.vue (hunk 12 above)
Comment (AuroraHuang22) [reply layouts/wallet.vue]: [🥅 Throw connection error](https://github.com/likecoin/app-like-co/pull/474/commits/cc9826077540dfdcd12ebe0f57470e7419f8c922)

File: components/AppHeader.vue (hunk 9 above)
Comment (AuroraHuang22) [reply components/AppHeader.vue]: [🥅 Throw connection error](https://github.com/likecoin/app-like-co/pull/474/commits/cc9826077540dfdcd12ebe0f57470e7419f8c922)

File: components/AppHeader.vue (hunk 8 above)
Comment (AuroraHuang22) [reply components/AppHeader.vue]: [🥅 Throw connection error](https://github.com/likecoin/app-like-co/pull/474/commits/cc9826077540dfdcd12ebe0f57470e7419f8c922)

File: layouts/wallet.vue (hunk 10 above)
Comment (AuroraHuang22) [reply layouts/wallet.vue]: [🎨 Add constant for authorization permission](https://github.com/likecoin/app-like-co/pull/474/commits/5c2c13370611651bc56f227ddcdf07466b7ea02a)

File: components/AppHeader.vue (hunk 13 above)
Comment (AuroraHuang22) [reply components/AppHeader.vue]: [🎨 Add constant for authorization permission](https://github.com/likecoin/app-like-co/pull/474/commits/5c2c13370611651bc56f227ddcdf07466b7ea02a)

File: pages/auth/redirect.vue (hunk 14 above)
Comment (AuroraHuang22) [reply pages/auth/redirect.vue]: [🎨 Add constant for authorization permission](https://github.com/likecoin/app-like-co/pull/474/commits/5c2c13370611651bc56f227ddcdf07466b7ea02a)

File: store/wallet.ts (hunk 17 above)
Comment (AuroraHuang22) [reply store/wallet.ts]: [🔥 Remove unused eslint-disable comments](https://github.com/likecoin/app-like-co/pull/474/commits/3c503f279abf7736b110d71552bdd0c5ac72018e)

File: store/book-api.ts (hunk 18 above)
Comment (AuroraHuang22) [reply store/book-api.ts]: [🥅 Throw error with axios response data](https://github.com/likecoin/app-like-co/pull/474/commits/11d36aca5a66a60714fe5a71bd340513cb7fad3b)

**Review by williamchong — APPROVED** [review]

**Pull request merged**